### ISSUE TITLE
Feature request: Generic types for query functions

### ISSUE THREAD

**Issue by renet** [KonnorRogers/shadow-dom-testing-library#52]
The `@testing-library/dom` uses generics for [typing the return value of their query functions](https://github.com/testing-library/dom-testing-library/blob/main/types/get-queries-for-element.d.ts) like so:

```ts
getByText<T extends HTMLElement = HTMLElement>(
  ...args: Parameters<BoundFunction<queries.GetByText<T>>>
): ReturnType<queries.GetByText<T>>
```

The `shadow-dom-testing-library` however only returns an element of type `HTMLElement` without the possibility to set a different type. It would be great, if one had the possibility to set a custom type when using a shadow query function like so:

```ts
screen.getByShadowText<HTMLButtonElement>("Submit")
```

**Comment by KonnorRogers** [comment]
@renet Thanks, I'll see what I can do, would definitely be a nice quality of life improvement. I _think_ it should be doable.

**Comment by KonnorRogers** [comment]
Implemented in v1.11.1

Let me know if it works for you @renet !

**Comment by renet** [comment]
Thank you very much! 💛 Works like a charm with regular selectors, but sub-selectors of `within(...)` are still not typable.

**Comment by KonnorRogers** [comment]
@renet Totally annoyed with how RTL doesn't preserve query function signatures, but I managed to get it working with a fantastically ugly hack.

**Comment by KonnorRogers** [comment]
Released in v1.11.2
